docs: add missing locators + improvements (#18404)
microsoft/playwright: docs/src/locators.md

[Locator]s are the central piece of Playwright's auto-waiting and retry-ability. In a nutshell, locators represent
a way to find element(s) on the page at any moment.

### Quick Guide

These are the recommended built in locators.

- [`method: Page.getByRole`] to locate by [ARIA role](https://www.w3.org/TR/wai-aria-1.2/#roles), [ARIA attributes](https://www.w3.org/TR/wai-aria-1.2/#aria-attributes) and [accessible name](https://w3c.github.io/accname/#dfn-accessible-name).
- [`method: Page.getByText`] to locate by text content.

> - [`method: Page.getByLabel`] to locate a form control by associated label's text.

Reviewer: Should we add links to the sections below for easy navigation? Not feeling strongly, just asking.

- [`method: Page.getByPlaceholder`] to locate an input by placeholder.
- [`method: Page.getByAltText`] to locate an element, usually image, by its text alternative.
- [`method: Page.getByTitle`] to locate an element by its title.
- [`method: Page.getByTestId`] to locate an element based on its `data-testid` attribute (other attribute can be configured).

```js
await page.getByLabel('User Name').fill('John');

await page.getByLabel('Password').fill('secret-password');

await page.getByRole('button', { name: 'Sign in' }).click();

await expect(page.getByText('Welcome, John!')).toBeVisible();
```

```java
page.getByLabel("User Name").fill("John");

page.getByLabel("Password").fill("secret-password");

page.getByRole("button", { name: "Sign in" }).click();

assertThat(page.getByText("Welcome, John!")).toBeVisible();
```

```python async
await page.get_by_label("User Name").fill("John")

await page.get_by_label("Password").fill("secret-password")

await page.get_by_role("button", { name: "Sign in" }).click()

> await page.get_by_text("Welcome, John!").toBeVisible();

Reviewer: Missing `expect`

```

```python sync
page.get_by_label("User Name").fill("John")

page.get_by_label("Password").fill("secret-password")

page.get_by_role("button", { name: "Sign in" }).click()

> page.get_by_text("Welcome, John!").toBeVisible();

Reviewer: Missing `expect`

```

```csharp
await page.GetByLabel('User Name').FillAsync('John');

await page.GetByLabel('Password').FillAsync('secret-password');

await page.GetByRole('button', { name: 'Sign in' }).ClickAsync();

> await expect(page.GetByText('Welcome, John!')).ToBeVisibleAsync();

Reviewer: Wrong syntax.

[Locator]s are the central piece of Playwright's auto-waiting and retry-ability. In a nutshell, locators represent
a way to find element(s) on the page at any moment.

### Quick Guide

These are the recommended built in locators and are available since version 1.27.

> - `page.getByRole` to locate by [ARIA role](https://www.w3.org/TR/wai-aria-1.2/#roles), [ARIA attributes](https://www.w3.org/TR/wai-aria-1.2/#aria-attributes) and [accessible name](https://w3c.github.io/accname/#dfn-accessible-name).

Author: damn. its so ugly but ok I get it

```python async
await page.get_by_role("button", name=re.compile("submit", re.IGNORECASE)).click()

await page.get_by_role("checkbox", checked=True, name="Check me"]).check()
```

```python sync
page.get_by_role("button", name=re.compile("submit", re.IGNORECASE)).click()

page.get_by_role("checkbox", checked=True, name="Check me"]).check()
```

```java
page.getByRole("button", new Page.GetByRoleOptions().setName(Pattern.compile("submit", Pattern.CASE_INSENSITIVE))).click();

page.getByRole("checkbox", new Page.GetByRoleOptions().setChecked(true).setName("Check me"))).check();
```

```csharp
await page.GetByRole("button", new() { Name = new Regex("submit", RegexOptions.IgnoreCase) }).ClickAsync();

await page.GetByRole("checkbox", new() { Checked = true, Name = "Check me" }).CheckAsync();
```

Role locators follow W3C specificaitons for [ARIA role](https://www.w3.org/TR/wai-aria-1.2/#roles), [ARIA attributes](https://www.w3.org/TR/wai-aria-1.2/#aria-attributes) and [accessible name](https://w3c.github.io/accname/#dfn-accessible-name).

Note that role locators **do not replace** accessibility audits and conformance tests, but rather give early feedback about the ARIA guidelines.

> ### Locate by label text with getByLabel

Author: Aahh, now I understand why you do that. its just really ugly in the sidebar having the method there. so perhaps we can just drop it and leave as 
### Locate by label text

Most form controls usually have dedicated labels that could be conveniently used to interact with the form. In this case, you can locate the control by its associated label using [`method: Page.getByLabel`].

For example, consider the following DOM structure.

```html
<label for="password">Password:</label><input type="password">
```

You can fill the input after locating it by the label text:

```js
await page.getByLabel('Password').fill('secret');
```

```java
page.getByLabel("Password").fill("secret");
```

```python async
await page.get_by_label("Password").fill("secret")
```

```python sync
page.get_by_label("Password").fill("secret")
```

```csharp
await page.GetByLabel("Password").FillAsync("secret");
```

### Locate by placeholder text with getByPlaceholder

> [`method: Page.getByPlaceholder`]

Reviewer: This method link is out of place.

> Inputs may have a placeholder attribute to hint to the user what value should be entered. You can locate the placeholder using [`method: Page.getByPlaceholder`].

Reviewer: ```suggestion
Inputs may have a placeholder attribute to hint to the user what value should be entered. You can locate such an input using [`method: Page.getByPlaceholder`].
```

For example, consider the following DOM structure.

```html
 <input id="email" name="email" type="email" placeholder="name@example.com">
```

You can fill the input after locating it by the label text:

```js
await page.getByPlaceholder("name@example.com").fill("playwright@microsoft.com");
```

```java
page.getByPlaceholder("name@example.com").fill("playwright@microsoft.com");
```

```python async
await page.get_by_placeholder("name@example.com").fill("playwright@microsoft.com")
```

```python sync
page.get_by_placeholder("name@example.com").fill("playwright@microsoft.com")
```

```csharp
await page.GetByPlacheolder("name@example.com").FillAsync("playwright@microsoft.com");
```

> ### Locate by text using getByText

Author: I think you should use getByRole before you use a getByText and getByTestId should be your last resort if you cannot get by a role or by a text. But thats just my opinion and Kent's opinion too. It's also how testing library layout their docs

> The easiest way to find an element is to look for the text it contains. You can match by a substring, exact string, or a regular expression when using [`method: Page.getByText`]

Reviewer: nit: missing punctuation at the end.

Reviewer: Still missing :)

Matching by text always normalizes whitespace, even with exact match. For example, it turns multiple spaces into one, turns line breaks into spaces and ignores leading and trailing whitespace.
:::

### Locate by alt text with getByAltText

> All images should have a valid alt tag which describes the image. You can locate the placeholder using [`method: Page.getByPlaceholder`].

Reviewer: ```suggestion
All images should have an `alt` attribute that describes the image. You can locate an image based on the text alternative using [`method: Page.getByAltText`].
```

For example, consider the following DOM structure.

```html
<img alt="playwright logo" src="/playwright-logo.png" />
```

> You can fill the input after locating it by the label text:

Reviewer: ```suggestion
You can click on the image after locating it by the text alternative:
```

```js
await page.getByAltText('playwright logo').click();
```

```java
page.getByAltText("playwright logo").click();
```

```python async
await page.get_by_alt_text("playwright logo").click()
```

```python sync
page.get_by_alt_text("playwright logo").click()
```

```csharp

> await page.getByAltText("playwright logo").click();

Reviewer: ```suggestion
await page.GetByAltText("playwright logo").ClickAsync();
```

```
### Locate by title with getByTitle

> [`method: Page.getByTitle`]

Reviewer: This method link is out of place.

Locate an element with a matching title attribute using [`method: Page.getByTitle`].

For example, consider the following DOM structure.

```html

> <span title='Close'></span>

Reviewer: I think the better example would be:

```html
<span title='Issues count'>25 issues</span>
```

```js
await expect(page.getByTitle('Issues count')).toHaveText('25 issues');
```

```

You can fill the input after locating it by the label text:

```js
await frame.getByTitle('Close').click();
```

```java
page.getByTitle("Close").click();
```

```python async
await page.get_by_label("Close").click()
```

```python sync
page.get_by_label("Close").click()
```

```csharp
await page.GetByTitle("Close").Click();
```

### Define explicit contract and use getByTestId

> User-facing attributes like text or accessible name can change over time. In this case it is convenient to define explicit test ids using [`method: Page.getByTestId`].

Reviewer: You are not defining test ids using this method, you only query using this method. Perhaps we should expand on "how to define test ids" to make this clear?

Matching by text always normalizes whitespace, even with exact match. For example, it turns multiple spaces into one, turns line breaks into spaces and ignores leading and trailing whitespace.
:::

### Locate by alt text

All images should have an `alt` attribute that describes the image. You can locate an image based on the text alternative using [`method: Page.getByAltText`].

For example, consider the following DOM structure.

```html
<img alt="playwright logo" src="/playwright-logo.png" />
```

You can click on the image after locating it by the text alternative:

```js
await page.getByAltText('playwright logo').click();
```

```java
page.getByAltText("playwright logo").click();
```

```python async
await page.get_by_alt_text("playwright logo").click()
```

```python sync
page.get_by_alt_text("playwright logo").click()
```

```csharp
await page.GetByAltText("playwright logo").ClickAsync();
```
### Locate by title

Locate an element with a matching title attribute using [`method: Page.getByTitle`].

For example, consider the following DOM structure.

```html
<span title='Issues count'>25 issues</span>
```

> You can fill the input after locating it by the label text:

Reviewer: This text is about something else.

```js

> await frame.getByTitle('Issues count').toHaveText('25 issues');

Reviewer: ```suggestion
await expect(page.getByTitle('Issues count')).toHaveText('25 issues');
```

Same for other languages.

page.getByRole("section").filter(new Locator.FilterOptions().setHas(page.getByTestId("subscribe-button")))
```
```python async

> page.get_by_role("section"), has=page.get_by_test_id("subscribe-button")

Reviewer: ```suggestion
page.get_by_role("section").filter(has=page.get_by_test_id("subscribe-button"))
```

```
```python sync

> page.get_by_role("section"), has=page.get_by_test_id("subscribe-button")

Reviewer: ```suggestion
page.get_by_role("section").filter(has=page.get_by_test_id("subscribe-button"))
```
